Monitoring as code foundation page (#2971)
sensu/sensu-docs: content/sensu-go/6.2/operations/monitoring-as-code/_index.md

---
title: "Monitoring as Code with Sensu"
linkTitle: "Monitoring as Code"
description: "Read this page to learn about following a monitoring-as-code approach with Sensu."
product: "Sensu Go"
version: "6.2"
weight: 5
layout: "single"
toc: true
menu:
  sensu-go-6.2:
    parent: operations
    identifier: monitoring-as-code
---

Sensu supports end-to-end monitoring as code so you can manage your observability and monitoring workflows the same way you build, test, and deploy your applications and infrastructure.

> Sensu allows you to define your entire monitoring workflow as declarative YAML or JSON code in configuration files, from collecting observability events and metrics and diagnosing issues through sending alerts and automatically remediating problems.

Reviewer: ```suggestion
Through the combination of composable building blocks and robust APIs, Sensu allows you to define your entire monitoring workflow as declarative YAML or JSON code in configuration files. From collecting observability events and metrics and diagnosing issues through sending alerts and automatically remediating problems, monitoring workflows can be shared and remixed across your organization.
```

In the monitoring as code approach, when a new endpoint starts up, like a cloud compute instance or Kubernetes Pod, Sensu automatically registers itself with the platform and starts collecting monitoring and observability data according to the code in your configuration files.

> If you manage your monitoring and observability workflow in the same way as the systems you're monitoring &mdash; as code, using a centralized continuous integration/continuous delivery (CI/CD) pipeline &mdash; you can align your monitoring with your product and services and improve visibility, reliability, and repeatability.

Reviewer: ```suggestion
If you manage your monitoring and observability workflow in the same way you manage the systems you're monitoring &mdash; as code, using a centralized continuous integration/continuous delivery (CI/CD) pipeline &mdash; you can align your monitoring with your product and services and improve visibility, reliability, and repeatability.
```

- Your entire team can share, edit, review, and version Sensu configuration files just like they would with other "as-code" solutions.
- Export your Sensu configuration from one environment and initialize another environment with the same configuration.
- Use Sensu configuration files to remove, restore, back up, and recover Sensu instances.

> 

Author: Are there other monitoring as code benefits to list here?

Author: I should use the "portable and repeatable" phrasing

## Create a monitoring as code repository

You can use any source control repository for maintaining your monitoring workflow configuration files.

> Your monitoring as code repository should include configuration files with the resource definitions you use in your monitoring workflow.

Author: I want to suggest what these configuration files might look like, but I'm not sure what we'd recommend.

For example, would we suggest putting all resources for a particular configuration into one file, like `test_resources.yaml`? Or would each file contain a type of resource, like `test_handlers.yaml`, `test_checks.yaml`, and so on?

Or would it be something else completely?

Reviewer: Sensu Flow [readme](https://github.com/sensu/sensu-flow#how-it-works) prescribes a directory structure like this:

```
.sensu/
  cluster/
    namespaces.yml
  namespaces/
    <namespace>/
      checks/
      hooks/
      filters/
      handlers/
      handelersets/
      mutators/
```

For Sensu Flow specifically, I don't think single resource per file vs. multiple resources per file is a concern.

More generally, I think it depends on the context. If I'm sharing a complete, end-to-end Sensu workflow with a colleague, I might put that all in one file for ease of reading. In the context of a code repository I am more likely to reuse an asset like the PagerDuty handler in multiple workflows, so storing each resource in an individual file would probably make more sense.

> Use `sensuctl dump` to [export some or all of your existing Sensu resource definitions][7] to standard out (STDOUT) or to a file.

Author: We talked about using "export" and "initialize" instead of "backup" and "recover" for this scenario.

We could:
- Reword and rename https://docs.sensu.io/sensu-go/latest/sensuctl/back-up-recover/ to use "export" and "restore" and "initialize"
- Recreate a doc similar to https://docs.sensu.io/sensu-go/latest/sensuctl/back-up-recover/ but focused on the monitoring as code use case (exporting and reinitializing)

I think the second option is better. I'm not sure if it needs to be a separate document inside a "Monitoring as Code" sub-category or if it would be brief enough to include here on this page.

Author: I've been thinking about this some more. I think the best option is including an "export and reproduce" section here on the MaC page.

spec:
  check_hooks: null
  command: check-cpu.rb -w 75 -c 90
  env_vars: null
  handlers:
  - slack
  high_flap_threshold: 0
  interval: 60
  low_flap_threshold: 0
  output_metric_format: ""
  output_metric_handlers: null
  proxy_entity_name: ""
  publish: true
  round_robin: false
  runtime_assets:
  - cpu-checks-plugins
  - sensu-ruby-runtime
  secrets: null
  stdin: false
  subdue: null
  subscriptions:
  - system
  timeout: 0
  ttl: 0
{{< /code >}}

{{< code shell "JSON" >}}
sensuctl check info check-cpu --format json
{
  "command": "check-cpu.rb -w 75 -c 90",
  "handlers": [
    "slack"
  ],
  "high_flap_threshold": 0,
  "interval": 60,
  "low_flap_threshold": 0,
  "publish": true,
  "runtime_assets": [
    "cpu-checks-plugins",
    "sensu-ruby-runtime"
  ],
  "subscriptions": [
    "system"
  ],
  "proxy_entity_name": "",
  "check_hooks": null,
  "stdin": false,
  "subdue": null,
  "ttl": 0,
  "timeout": 0,
  "round_robin": false,
  "output_metric_format": "",
  "output_metric_handlers": null,
  "env_vars": null,
  "metadata": {
    "name": "check-cpu",
    "namespace": "default",
    "created_by": "admin"
  },
  "secrets": null
}
{{< /code >}}

{{< /language-toggle >}}

> Many of our [guides][6] demonstrate how to create, update, and retrieve resource definitions in this way.

Author: They *will* include this instruction, although they mostly don't right now. See https://github.com/sensu/sensu-docs/pull/2962 -- we think the monitoring as code project should include reworking pages that use mainly sensuctl commands to create resources so that they instead focus on complete code examples.

I also think we need to expand this explanation so it's clear to users what they should do with the resource definitions once they get them with `sensuctl [resource] info [resource name]` -- for example, save them in a file like `checks.yaml`?

Reviewer: To capture resource to a new file (or overwrite existing), e.g. resource.yml:
`sensuctl [resource] info [resource name] > resource.yml`

To capture resource to a new file (or overwrite existing), e.g. resource.yml and see output on standard out:
`sensuctl [resource] info [resource name] | tee resource.yml`

To append resource to an existing file:
`sensuctl [resource] info [resource name] >> resource.yml`

To append resource to an existing file and show the output on standard out:
`sensuctl [resource] info [resource name] | tee -a resource.yml`

## Implement a monitoring as code workflow

There's no one "correct" way to implement monitoring as code.

> To get started, try [SensuFlow][5], our git-based approach to managing Sensu resources.

Reviewer: ```suggestion
To get started, try [SensuFlow][5], our GitHub Action for managing Sensu resources via repository commits.
```

https://github.com/marketplace/actions/sensuflow might be a better link than the github repository, as the marketplace page gives folks an easy copy/paste to get started using the action with their repository

> 

Author: Is there more to say about this? There must be common requirements, like the four things mentioned for SensuFlow (an RBAC service account, code repo, labeling convention, and CI/CD integration).

Anything else we should say here?

Reviewer: Even without a CI/CD pipeline in place, monitoring as code can make repetitive Sensu administration tasks easier, e.g. using revision control ensures you have a change history, and the sensuctl prune command can be used to apply changes in your repository in a consistent way using a single command. Once you're ready to expand your monitoring as code practices to involve CI/CD and multiple Sensu clusters, you'll want to bootstrap your Sensu clusters with service accounts that are granted permissions for managing Sensu configuration. We can provide a couple examples, e.g. a role and role binding that allows its subjects to manage all configuration in a given namespace.

Labeling conventions are definitely something we need to cover here, but I'm not clear on how this works at the moment.

## Best practices for monitoring as code

The repository of resource definitions can be any manner of source control repository (git, subversion, etc.). While a specific directory structure is not required, we will be suggesting one later in this document.

- To maintain consistency, save all of your resources as only one file type: YAML or JSON.
- Include all dependencies within a resource definition.

> For example, if a handler requires a dynamic runtime asset and a secret, include the asset and secret definitions with the definition for the handler itself.

Author: These are cabbaged from Jef and Todd's blog post about SensuFlow. Are there other best practices to mention?

Maybe this section should include the content I currently have above under "Implement a monitoring as code workflow"

## Monitoring as code with SensuFlow

[SensuFlow][5] is a git-based, prescriptive monitoring as code workflow that uses [sensuctl][2] (including [sensuctl prune][3]) to synchronize your monitoring and observability code with your Sensu deployments.

{{% notice note %}}
**NOTE**: SensuFlow is available for technical preview, and individual components in the workflow may change.
Before you use SensuFlow in production, test it in a development environment or a dedicated test namespace in your current environment.
{{% /notice %}}

SensuFlow requires:

- A monitoring code repository of Sensu resource definitions
- A Sensu [role-based access control (RBAC)][4] service account with permission to manage all resources in your repository
- A resource labeling convention to designate which resources the SensuFlow workflow should manage
- Integration with your CI/CD system to run sensuctl commands as the service account user from the repository of resource definitions

Use the [SensuFlow GitHub Action][1], a turnkey reference implementation, to create your own monitoring as code workflow.

> 

Author: I did not add more in this section because I supposed the SensuFlow GitHub Actions instructions would live in the repo README, which makes sense to me.

But is there more to say here to help users be successful with trying out SensuFlow?

Author: [Slack comment from Caleb](https://sensu.slack.com/archives/C6AF34L6R/p1611871786134400)

Author: The published blog post: https://sensu.io/blog/monitoring-as-code-with-sensu-flow

> [1]: https://github.com/sensu/sensu-flow

Author: ```suggestion
[1]: https://github.com/marketplace/actions/sensuflow
```

- Share, edit, review, and version your Sensu workflow configuration files just like you would with other "as code" solutions, within one team or among teams across your organization.
- Maintain revision control and change history for your monitoring workflows.
- Export your Sensu configuration from one environment and initialize another environment with the same configuration.
- Remove, restore, back up, and recover Sensu instances based on Sensu workflow configuration files.
- Use `sensuctl prune` to consistently apply changes to your monitoring workflows with a single command.
- Include your monitoring workflow in your centralized continuous integration/continuous delivery (CI/CD) pipeline to keep your monitoring aligned with your product and services.

## Create a monitoring as code repository

A monitoring as code repository will include configuration files that contain the resource definitions you use in your monitoring workflow.
You can use any source control repository for maintaining your monitoring workflow configuration files.

> 

Author: Maybe this should say something like "To get started with monitoring as code, you'll need configuration files that contain your resource definitions, a repository where you can store and maintain your configuration files, and a file structure and naming convention."

Before you use SensuFlow in production, test it in a development environment or a dedicated test namespace in your current environment.
{{% /notice %}}

SensuFlow requires:

- A monitoring code repository of Sensu resource definitions
- A Sensu [role-based access control (RBAC)][4] service account with permission to manage all resources in your repository
- A resource labeling convention to designate which resources the SensuFlow workflow should manage
- Integration with your CI/CD system to run sensuctl commands as the service account user from the repository of resource definitions

Read the [SensuFlow GitHub Action marketplace page][1] and [Monitoring as code with Sensu Go and SensuFlow][12] for more information.

## Use CI/CD with monitoring as code

When you're ready to expand your monitoring as code practices to involve CI/CD and multiple Sensu clusters, you can bootstrap your Sensu clusters with service accounts that are granted permissions for managing Sensu configuration.

For example, the following role and role binding configuriations allow subjects to manage all configuration in a given namespace:

> TODO: add role and role binding examples

Author: Would this be the same thing I started trying to explain in https://github.com/sensu/sensu-docs/pull/2935/files?

Author: This section needs a bit more explanation about how CI/CD would benefit a monitoring as code approach.

- Maintain revision control and change history for your monitoring workflows.
- Export your Sensu configuration from one environment and initialize another environment with the same configuration.
- Remove, restore, back up, and recover Sensu instances based on Sensu workflow configuration files.
- Use `sensuctl prune` to consistently apply changes to your monitoring workflows with a single command.
- Include your monitoring workflow in your centralized continuous integration/continuous delivery (CI/CD) pipeline to keep your monitoring aligned with your product and services.

To get started with monitoring as code, you'll need a [repository][11] and [configuration files][13] that contain your resource definitions.

## Create a monitoring as code repository

A monitoring as code repository will include configuration files that contain the resource definitions you use in your monitoring workflow.
You can use any source control repository for maintaining your monitoring workflow configuration files, but you'll need to decide how to structure your repository.

> The way you will use your Sensu resource [configuration files][13] will help you choose the best structure for your monitoring as code repository:
> 
> - If you want to share complete end-to-end Sensu workflows with your colleagues, you might save all of the resource definitions for each workflow in a single configuration file.
> This allows others to read through an entire workflow without interruption.
> - If you are more likely to share individual workflow components, it makes more sense to use individual configuration files for different types of resources, like one file for all of your checks, one file for all of your handlers, and so on.
> - If you want to facilitate even more granular sharing, you can save one resource definition per file.

Author: ```suggestion
The way you will use your Sensu resource [configuration files][13] will help you choose the best structure for your monitoring as code repository.

For example, if you are likely to share individual workflow components or manage your Sensu configuration files as part of your CI/CD workflow, it probably makes sense to use individual configuration files for different types of resources: one file for all of your checks, one file for all of your handlers, and so on.
If you want to facilitate even more granular sharing, you can save one resource definition per file.

If you want to share complete end-to-end Sensu workflows with your colleagues, you might save all of the resource definitions for each workflow in a single configuration file.
This allows others to read through an entire workflow without interruption, and it's convenient for demonstrating a complete Sensu workflow.
However, a single configuration file that includes every resource type isn't the best structure for CI/CD management or sharing resources among teams.
```

> 

Author: Just occurred to me -- maybe we want to say something about considering what's best if you're going to eventually manage these via CI/CD.

The way you will use your Sensu resource [configuration files][13] will help you choose the best structure for your monitoring as code repository.

For example, if you are likely to share individual workflow components or manage your Sensu configuration files as part of your CI/CD workflow, it probably makes sense to use individual configuration files for different types of resources: one file for all of your checks, one file for all of your handlers, and so on.
If you want to facilitate even more granular sharing, you can save one resource definition per file.

If you want to share complete end-to-end Sensu workflows with your colleagues, you might save all of the resource definitions for each workflow in a single configuration file.
This allows others to read through an entire workflow without interruption, and it's convenient for demonstrating a complete Sensu workflow.
However, a single configuration file that includes every resource type isn't the best structure for CI/CD management or sharing resources among teams.

> For example, [SensuFlow][5], our GitHub Action for managing Sensu resources via repository commits, requires a repository structure organized by clusters and namespaces.

Author: ```suggestion
[SensuFlow][5], our GitHub Action for managing Sensu resources via repository commits, requires a repository structure organized by clusters and namespaces.
```
